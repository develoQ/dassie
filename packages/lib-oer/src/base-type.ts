import { ParseError, SerializeError } from "./utils/errors"
import type { ParseContext, SerializeContext } from "./utils/parse"
import { TagClass, TagMarker, tagClassMarkerMap } from "./utils/tag"

export type AnyOerType = OerType<unknown>

export type Infer<TOerType extends AnyOerType> = TOerType extends OerType<
  infer T
>
  ? T
  : never

export interface ParseOptions {
  allowNoncanonical?: boolean
}

export type IntermediateSerializationResult = readonly [
  serialize: (
    context: SerializeContext,
    offset: number
    // eslint-disable-next-line @typescript-eslint/no-invalid-void-type
  ) => SerializeError | void,
  length: number
]

export abstract class OerType<TValue> {
  _tag: readonly [tagValue: number, tagClass: TagMarker] | undefined
  _optional = false

  abstract parseWithContext(
    context: ParseContext,
    offset: number
  ): readonly [value: TValue, length: number] | ParseError
  abstract serializeWithContext(
    value: TValue
  ): IntermediateSerializationResult | SerializeError

  parse(
    input: Uint8Array,
    offset = 0,
    options: ParseOptions = {}
  ):
    | { success: true; value: TValue; length: number }
    | { success: false; failure: ParseError } {
    const context: ParseContext = {
      uint8Array: input,
      dataView: new DataView(input.buffer),
      allowNoncanonical: options.allowNoncanonical ?? false,
    }
    const result = this.parseWithContext(context, offset)

    if (result instanceof ParseError) {
      return { success: false, failure: result }
    }

    if (!context.allowNoncanonical && result[1] !== input.byteLength - offset) {
      return {
        success: false,
        failure: new ParseError(
          "non-canonical encoding - additional bytes present after the expected end of data",
          input,
          result[1]
        ),
      }
    }

    return { success: true, value: result[0], length: result[1] }
  }

  serialize(
    value: TValue
  ):
    | { success: true; value: Uint8Array }
    | { success: false; failure: SerializeError } {
    const prediction = this.serializeWithContext(value)

    if (prediction instanceof SerializeError) {
      return { success: false, failure: prediction }
    }

    const [serialize, length] = prediction

    const buffer = new Uint8Array(length)

    const result = serialize(
      {
        uint8Array: buffer,
        dataView: new DataView(buffer.buffer),
      },
      0
    )

    return result instanceof SerializeError
      ? { success: false, failure: result }
      : { success: true, value: buffer }
  }

  tag(
    tagValue: number,
    tagClass: Exclude<TagClass, "universal"> = "context"
  ): this {
    this._tag = [tagValue, tagClassMarkerMap[tagClass]] as const
    return this
  }

  optional(): this {
    this._optional = true
    return this
  }
}
