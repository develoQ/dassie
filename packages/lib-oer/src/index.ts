export type { Infer, OerType, AnyOerType, ParseOptions } from "./base-type"
export * from "./bitstring"
export * from "./boolean"
export * from "./choice"
export * from "./enumerated"
export * from "./integer-as-bigint"
export * from "./integer-as-number"
export * from "./octet-string"
export * from "./sequence-of"
export * from "./sequence"
export * from "./string"
