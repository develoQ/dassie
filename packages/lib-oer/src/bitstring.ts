import type { ReadonlyTuple } from "type-fest"

import { OerType } from "./base-type"
import { ParseError, SerializeError } from "./utils/errors"
import {
  parseLengthPrefix,
  predictLengthPrefixLength,
  serializeLengthPrefix,
} from "./utils/length-prefix"
import type { BaseContext, ParseContext, SerializeContext } from "./utils/parse"

type InferBitstringValue<T extends number> = ReadonlyTuple<boolean, T>

interface BitstringOptions {
  variableLength?: boolean
}

export const parseBitstring = (
  context: ParseContext,
  offset: number,
  size: number
) => {
  const byteLength = Math.ceil(size / 8)
  const output = Array.from({ length: size })
  for (let byteIndex = 0; byteIndex < byteLength; byteIndex++) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const byte = context.uint8Array[offset + byteIndex]!

    for (let bitIndex = 0; bitIndex < 8; bitIndex++) {
      if (byteIndex * 8 + bitIndex >= size) {
        break
      }
      const bit = (byte >> (7 - bitIndex)) & 1
      output[byteIndex * 8 + bitIndex] = bit === 1
    }
  }
  return output
}

export const serializeBitstring = (
  context: BaseContext,
  offset: number,
  input: InferBitstringValue<number>
) => {
  const { uint8Array } = context
  const inputAsArray = input as boolean[]
  const byteLength = Math.ceil(inputAsArray.length / 8)

  for (let byteIndex = 0; byteIndex < byteLength; byteIndex++) {
    let byte = 0
    for (let bitIndex = 0; bitIndex < 8; bitIndex++) {
      const index = byteIndex * 8 + bitIndex
      if (index >= inputAsArray.length) break
      byte |= inputAsArray[index] ? 1 << (7 - bitIndex) : 0
    }
    uint8Array[offset + byteIndex] = byte
  }
}

export const bitstring = <TBitDefinition extends number>(
  bits: TBitDefinition,
  { variableLength = false }: BitstringOptions = {}
) => {
  const OerVariableBitstring = class extends OerType<
    InferBitstringValue<TBitDefinition>
  > {
    parseWithContext(context: ParseContext, offset: number) {
      const { uint8Array } = context
      const result = parseLengthPrefix(context, offset)

      if (result instanceof ParseError) {
        return result
      }

      const [length, lengthOfLength] = result

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const unusedBits = uint8Array[offset + lengthOfLength]!

      if (unusedBits > 7) {
        return new ParseError(
          "unable to read bitstring - unused bits greater than 7",
          uint8Array,
          offset + 1
        )
      }

      const output = parseBitstring(
        context,
        offset + lengthOfLength + 1,
        (length - 1) * 8 - unusedBits
      )

      return [
        output as InferBitstringValue<TBitDefinition>,
        lengthOfLength + length,
      ] as const
    }

    serializeWithContext(input: ReadonlyTuple<boolean, TBitDefinition>) {
      const byteLength = Math.ceil(bits / 8)
      const unusedBits = 8 - (bits % 8)
      const length = byteLength + 1
      const lengthOfLength = predictLengthPrefixLength(length)
      if (lengthOfLength instanceof SerializeError) {
        return lengthOfLength
      }

      return [
        (context: SerializeContext, offset: number) => {
          const { uint8Array } = context
          const lengthOfLength = serializeLengthPrefix(
            length,
            uint8Array,
            offset
          )

          if (lengthOfLength instanceof SerializeError) {
            return lengthOfLength
          }

          uint8Array[offset + lengthOfLength] = unusedBits
          serializeBitstring(context, offset + lengthOfLength + 1, input)
          return
        },
        lengthOfLength + length,
      ] as const
    }
  }

  const OerFixedBitstring = class extends OerType<
    InferBitstringValue<TBitDefinition>
  > {
    parseWithContext(context: ParseContext, offset: number) {
      const byteLength = Math.ceil(bits / 8)

      if (context.uint8Array.length - offset < byteLength) {
        return new ParseError(
          "unable to read bitstring value - end of buffer",
          context.uint8Array,
          offset
        )
      }

      const output = parseBitstring(context, offset, bits)

      return [
        output as InferBitstringValue<TBitDefinition>,
        byteLength,
      ] as const
    }

    serializeWithContext(input: InferBitstringValue<TBitDefinition>) {
      const inputAsArray = input as boolean[]
      if (inputAsArray.length !== bits) {
        return new SerializeError(
          `unable to serialize bitstring - expected ${bits} bits, got ${inputAsArray.length}`
        )
      }

      return [
        (context: SerializeContext, offset: number) => {
          serializeBitstring(context, offset, input)
        },
        Math.ceil(bits / 8),
      ] as const
    }
  }
  return variableLength ? new OerVariableBitstring() : new OerFixedBitstring()
}
