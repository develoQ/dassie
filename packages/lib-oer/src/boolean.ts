import { OerType } from "./base-type"
import { ParseError } from "./utils/errors"
import type { ParseContext, SerializeContext } from "./utils/parse"

export const boolean = () => {
  const OerBoolean = class extends OerType<boolean> {
    parseWithContext(
      { uint8Array, allowNoncanonical }: ParseContext,
      offset: number
    ) {
      const value = uint8Array[offset]
      if (typeof value === "undefined") {
        return new ParseError(
          "unable to read boolean value - end of buffer",
          uint8Array,
          offset
        )
      }
      if (!allowNoncanonical && value !== 0 && value !== 0xff) {
        return new ParseError(
          "unable to read boolean value - non-canonical",
          uint8Array,
          offset
        )
      }
      return [!!value, 1] as const
    }
    serializeWithContext(input: boolean) {
      return [
        ({ uint8Array }: SerializeContext, offset: number) => {
          uint8Array[offset] = input ? 0xff : 0
        },
        1,
      ] as const
    }
  }
  return new OerBoolean()
}
