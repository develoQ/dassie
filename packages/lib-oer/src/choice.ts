import { AnyOerType, Infer, InferSerialize, OerType } from "./base-type"
import { ParseError, SerializeError } from "./utils/errors"
import type { ParseContext, SerializeContext } from "./utils/parse"
import {
  TAG_MARKER_CONTEXT,
  TagMarker,
  parseTag,
  predictTagLength,
  serializeTag,
  tagMarkerClassMap,
} from "./utils/tag"

/**
 * Change all properties of an interface to type never
 */
type Never<T> = { [P in keyof T]?: never }

/**
 * Returns a union of interfaces where each member is an object containing a single property which is of such type as inferred from the corresponding member of the input interface.
 */
type InferOptionValues<T extends Record<string, AnyOerType>> = {
  [K in keyof T]-?: { [P in K]: Infer<T[P]> } & Never<
    Pick<T, Exclude<keyof T, K>>
  >
}[keyof T]

type InferOptionSerializeValues<T extends Record<string, AnyOerType>> = {
  [K in keyof T]-?: { [P in K]: InferSerialize<T[P]> } & Never<
    Pick<T, Exclude<keyof T, K>>
  >
}[keyof T]

// Because JavaScript's bitwise operators are 32 bits and we are using two bits for the tag class, we only have 30 bits available to encode the tag value
// This should be plenty though since tag values are usually much smaller.
const MAX_SAFE_TAG_VALUE = Number.MAX_SAFE_INTEGER >>> 2

export const choice = <TOptions extends Record<string, AnyOerType>>(
  options: TOptions
) => {
  // Automatic tagging is enabled if none of the values are explicitly tagged
  const isAutomaticallyTagged = Object.values(options).every(
    (oer) => oer._tag == undefined
  )
  let index = 0
  // Tags are encoded here as an integer with the class marker in the lowest two bits and the tag value shifted two bits to the right.
  const keyToTagMap = new Map<keyof TOptions, number>()
  const tagToKeyMap = new Map<number, keyof TOptions>()

  for (const [key, oer] of Object.entries(options)) {
    const tag = isAutomaticallyTagged
      ? (index++ << 2) | TAG_MARKER_CONTEXT
      : oer._tag != undefined
      ? (oer._tag[0] << 2) | oer._tag[1]
      : undefined

    if (tag == undefined) {
      throw new Error(
        `OER choice: ${key} is not tagged and automatic tagging is disabled`
      )
    }

    if (tag > MAX_SAFE_TAG_VALUE) {
      throw new Error(
        `OER choice: ${key} has a tag value of ${tag} which is too large`
      )
    }

    if (tagToKeyMap.has(tag)) {
      throw new Error(
        `OER choice: ${key} has the same tag value as ${String(
          tagToKeyMap.get(tag)
        )}`
      )
    }

    keyToTagMap.set(key, tag)
    tagToKeyMap.set(tag, key)
  }

  Object.fromEntries(
    Object.entries(options).map(([key, oer]) => {
      return [
        key,
        isAutomaticallyTagged
          ? ([index++, TAG_MARKER_CONTEXT] as const)
          : oer._tag,
      ]
    })
  )

  const setHint = Object.entries(keyToTagMap)
    .map(([key, tag]) => `${key}(${String(tag)})`)
    .join(",")

  const OerChoice = class extends OerType<
    InferOptionValues<TOptions>,
    InferOptionSerializeValues<TOptions>
  > {
    parseWithContext(context: ParseContext, offset: number) {
      const tagResult = parseTag(context, offset)

      if (tagResult instanceof ParseError) {
        return tagResult
      }

      const [tag, tagClass, tagLength] = tagResult //?

      if (tag > MAX_SAFE_TAG_VALUE) {
        return new ParseError(
          `unable to parse choice value - tag value ${tag} is too large`,
          context.uint8Array,
          offset
        )
      }

      const key = tagToKeyMap.get((tag << 2) | tagClass)

      if (key == undefined) {
        return new ParseError(
          `unable to read choice value - tag ${tagMarkerClassMap[
            tagClass
          ].toUpperCase()} ${tag} not in set ${setHint}`,
          context.uint8Array,
          offset
        )
      }

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const oer = options[key]!

      const result = oer.parseWithContext(context, offset + tagLength)

      if (result instanceof ParseError) {
        return result
      }

      return [
        { [key]: result[0] } as InferOptionValues<TOptions>,
        tagLength + result[1],
      ] as const
    }

    serializeWithContext(input: InferOptionSerializeValues<TOptions>) {
      const key = Object.keys(input)[0]

      if (key == undefined) {
        return new SerializeError(
          `unable to serialize choice value - no option selected`
        )
      }

      const tag = keyToTagMap.get(key)

      if (tag == undefined) {
        return new SerializeError(
          `unable to serialize choice value - option ${key} not in set ${setHint}`
        )
      }

      const tagValue = tag >>> 2
      const tagClass = (tag & 3) as TagMarker

      const oer = options[key]

      if (oer == undefined) {
        return new SerializeError(
          `unable to serialize choice value - option ${key} not in set ${setHint}`
        )
      }

      const tagLength = predictTagLength(tagValue)

      if (tagLength instanceof SerializeError) {
        return tagLength
      }

      const valueResult = oer.serializeWithContext(input[key])

      if (valueResult instanceof SerializeError) {
        return valueResult
      }

      return [
        (context: SerializeContext, offset: number) => {
          const { uint8Array } = context

          const tagSerializeResult = serializeTag(
            tagValue,
            tagClass,
            uint8Array,
            offset
          )

          if (tagSerializeResult instanceof SerializeError) {
            return tagSerializeResult
          }

          const serializeResult = valueResult[0](context, offset + tagLength)

          if (serializeResult instanceof SerializeError) {
            return serializeResult
          }

          return
        },
        tagLength + valueResult[1],
      ] as const
    }
  }
  return new OerChoice()
}
