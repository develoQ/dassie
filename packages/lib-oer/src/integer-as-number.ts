import { OerType } from "./base-type"
import { ParseError, SerializeError } from "./utils/errors"
import type { ParseContext, SerializeContext } from "./utils/parse"
import { FixedRange, Range, parseRange } from "./utils/range"

export interface IntegerAsNumberOptions {
  range: FixedRange<number>
}

export const UINT_MIN_NUMBER = 0
export const UINT8_MAX_NUMBER = 255
export const UINT16_MAX_NUMBER = 65_535
export const UINT32_MAX_NUMBER = 4_294_967_295

export const INT8_MIN_NUMBER = -128
export const INT8_MAX_NUMBER = 127
export const INT16_MIN_NUMBER = -32_768
export const INT16_MAX_NUMBER = 32_767
export const INT32_MIN_NUMBER = -2_147_483_648
export const INT32_MAX_NUMBER = 2_147_483_647

export abstract class OerFixedIntegerNumber extends OerType<number> {
  abstract readonly type: "Uint" | "Int"
  abstract readonly size: 8 | 16 | 32

  constructor(readonly options: IntegerAsNumberOptions) {
    super()
  }

  parseWithContext({ uint8Array, dataView }: ParseContext, offset: number) {
    if (offset + this.size / 8 > dataView.byteLength) {
      return new ParseError(
        `unable to read fixed length integer of size ${
          this.size / 8
        } bytes - end of buffer`,
        uint8Array,
        uint8Array.byteLength
      )
    }

    const value = dataView[`get${this.type}${this.size}`](offset)

    if (value < this.options.range[0]) {
      return new ParseError(
        `unable to read fixed length integer of size ${
          this.size / 8
        } bytes - value ${value} is less than minimum value ${
          this.options.range[0]
        }`,
        uint8Array,
        offset
      )
    }

    if (value > this.options.range[1]) {
      return new ParseError(
        `unable to read fixed length integer of size ${
          this.size / 8
        } bytes - value ${value} is greater than maximum value ${
          this.options.range[1]
        }`,
        uint8Array,
        offset
      )
    }

    return [value, this.size / 8] as const
  }

  serializeWithContext(value: number) {
    const serializer = (context: SerializeContext, offset: number) => {
      if (value < this.options.range[0]) {
        return new SerializeError(`integer must be >= ${this.options.range[0]}`)
      }

      if (value > this.options.range[1]) {
        return new SerializeError(`integer must be <= ${this.options.range[1]}`)
      }

      context.dataView[`set${this.type}${this.size}`](offset, value)

      return
    }
    serializer.size = this.size / 8
    return serializer
  }
}

export class OerUint8Number extends OerFixedIntegerNumber {
  readonly size = 8
  readonly type = "Uint"
}
export class OerUint16Number extends OerFixedIntegerNumber {
  readonly size = 16
  readonly type = "Uint"
}
export class OerUint32Number extends OerFixedIntegerNumber {
  readonly size = 32
  readonly type = "Uint"
}

export class OerInt8Number extends OerFixedIntegerNumber {
  readonly size = 8
  readonly type = "Int"
}
export class OerInt16Number extends OerFixedIntegerNumber {
  readonly size = 16
  readonly type = "Int"
}
export class OerInt32Number extends OerFixedIntegerNumber {
  readonly size = 32
  readonly type = "Int"
}

export const integerAsNumber = (range: Range<number>) => {
  const [minimumValue, maximumValue] = parseRange(range)

  if (minimumValue == undefined || maximumValue == undefined) {
    throw new Error(
      "When using JavaScript numbers, a minimum and maximum must be provided. If your range is actually unbounded, use integerAsBigint instead."
    )
  }

  const fixedOptions: IntegerAsNumberOptions = {
    range: [minimumValue, maximumValue],
  }

  // Fixed size integer encodings
  if (minimumValue >= UINT_MIN_NUMBER) {
    if (maximumValue <= UINT8_MAX_NUMBER) {
      return new OerUint8Number(fixedOptions)
    } else if (maximumValue <= UINT16_MAX_NUMBER) {
      return new OerUint16Number(fixedOptions)
    } else if (maximumValue <= UINT32_MAX_NUMBER) {
      return new OerUint32Number(fixedOptions)
    }
  } else {
    if (minimumValue <= INT8_MIN_NUMBER && maximumValue <= INT8_MAX_NUMBER) {
      return new OerInt8Number(fixedOptions)
    } else if (
      minimumValue <= INT16_MIN_NUMBER &&
      maximumValue <= INT16_MAX_NUMBER
    ) {
      return new OerInt16Number(fixedOptions)
    } else if (
      minimumValue <= INT32_MIN_NUMBER &&
      maximumValue <= INT32_MAX_NUMBER
    ) {
      return new OerInt32Number(fixedOptions)
    }
  }

  throw new TypeError(
    "integerAsNumber only supports up to 32 bits due to JavaScript's limitations, please use integerAsBigint instead"
  )
}

export const uint8Number = () =>
  integerAsNumber([UINT_MIN_NUMBER, UINT8_MAX_NUMBER])
export const uint16Number = () =>
  integerAsNumber([UINT_MIN_NUMBER, UINT16_MAX_NUMBER])
export const uint32Number = () =>
  integerAsNumber([UINT_MIN_NUMBER, UINT32_MAX_NUMBER])
export const int8Number = () =>
  integerAsNumber([INT8_MIN_NUMBER, INT8_MAX_NUMBER])
export const int16Number = () =>
  integerAsNumber([INT16_MIN_NUMBER, INT16_MAX_NUMBER])
export const int32Number = () =>
  integerAsNumber([INT32_MIN_NUMBER, INT32_MAX_NUMBER])
