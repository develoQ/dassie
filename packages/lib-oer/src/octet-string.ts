import { OerType } from "./base-type"
import { ParseError, SerializeError } from "./utils/errors"
import {
  parseLengthPrefix,
  predictLengthPrefixLength,
  serializeLengthPrefix,
} from "./utils/length-prefix"
import {
  ParseContext,
  SafeUnsignedInteger,
  SerializeContext,
  isSafeUnsignedInteger,
} from "./utils/parse"
import { Range, parseRange } from "./utils/range"

export const OerFixedOctetString = class extends OerType<Uint8Array> {
  constructor(readonly length: SafeUnsignedInteger) {
    super()
  }

  parseWithContext({ uint8Array }: ParseContext, offset: number) {
    return [
      uint8Array.slice(offset, offset + this.length),
      this.length,
    ] as const
  }

  serializeWithContext(value: Uint8Array) {
    if (value.length !== this.length) {
      return new SerializeError(
        `Expected octet string of length ${this.length}, but got ${value.length}`
      )
    }

    return [
      (context: SerializeContext, offset: number) => {
        context.uint8Array.set(value, offset)
      },
      this.length,
    ] as const
  }
}

export const OerVariableOctetString = class extends OerType<Uint8Array> {
  constructor(
    readonly minimumLength?: SafeUnsignedInteger,
    readonly maximumLength?: SafeUnsignedInteger
  ) {
    super()
  }

  parseWithContext(context: ParseContext, offset: number) {
    const { uint8Array } = context
    const result = parseLengthPrefix(context, offset)

    if (result instanceof ParseError) {
      return result
    }

    const [length, lengthOfLength] = result

    return [
      uint8Array.slice(
        offset + lengthOfLength,
        offset + lengthOfLength + length
      ),
      lengthOfLength + length,
    ] as const
  }

  serializeWithContext(input: Uint8Array) {
    const length = input.length
    const lengthOfLengthPrefix = predictLengthPrefixLength(length)

    if (lengthOfLengthPrefix instanceof SerializeError) {
      return lengthOfLengthPrefix
    }

    return [
      ({ uint8Array }: SerializeContext, offset: number) => {
        const length = input.length
        const lengthOfLengthPrefix = serializeLengthPrefix(
          length,
          uint8Array,
          offset
        )

        if (lengthOfLengthPrefix instanceof SerializeError) {
          return lengthOfLengthPrefix
        }

        uint8Array.set(input, offset + lengthOfLengthPrefix)
        return
      },
      lengthOfLengthPrefix + length,
    ] as const
  }
}

export const octetString = (length?: Range<number>) => {
  const { minimum: minimumLength, maximum: maximumLength } = parseRange(length)
  if (minimumLength != undefined && !isSafeUnsignedInteger(minimumLength)) {
    throw new TypeError("minimumLength must be a safe unsigned integer")
  }
  if (maximumLength != undefined && !isSafeUnsignedInteger(maximumLength)) {
    throw new TypeError("maximumLength must be a safe unsigned integer")
  }

  if (
    minimumLength != undefined &&
    maximumLength != undefined &&
    minimumLength === maximumLength
  ) {
    return new OerFixedOctetString(minimumLength)
  }

  if (
    minimumLength != undefined &&
    maximumLength != undefined &&
    minimumLength > maximumLength
  ) {
    throw new TypeError(
      "minimumLength must be less than or equal to maximumLength"
    )
  }

  return new OerVariableOctetString(minimumLength, maximumLength)
}
