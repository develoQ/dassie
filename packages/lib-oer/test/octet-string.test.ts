import { describe, test } from "vitest"

import { octetString } from "../src/octet-string"
import { hexToUint8Array } from "../src/utils/hex"
import { parsedOk, serializedOk } from "./utils/result"
import { sampleBuffer } from "./utils/sample-buffer"

describe("octetString", () => {
  test("should be a function", ({ expect }) => {
    expect(octetString).toBeTypeOf("function")
  })

  describe("with fixed length", () => {
    const schema = octetString({ fixedLength: 22 })

    test("should return an object", ({ expect }) => {
      expect(schema).toBeTypeOf("object")
    })

    test("should serialize a valid value", ({ expect }) => {
      const value = schema.serialize(sampleBuffer.slice(0, 22))
      expect(value).toEqual(serializedOk(sampleBuffer.slice(0, 22)))
    })

    test("should parse a valid value", ({ expect }) => {
      const value = schema.parse(sampleBuffer.slice(0, 22))
      expect(value).toEqual(parsedOk(22, sampleBuffer.slice(0, 22)))
    })

    test("should refuse to serialize a value of the wrong length", ({
      expect,
    }) => {
      const value = schema.serialize(sampleBuffer.slice(0, 21))
      expect(value).toMatchSnapshot()
    })
  })

  describe("with variable length", () => {
    const schema = octetString()

    test("should return an object", ({ expect }) => {
      expect(schema).toBeTypeOf("object")
    })

    test("should serialize an empty buffer", ({ expect }) => {
      const value = schema.serialize(new Uint8Array(0))
      expect(value).toEqual(serializedOk("00"))
    })

    test("should serialize a buffer with one byte", ({ expect }) => {
      const value = schema.serialize(hexToUint8Array("12"))
      expect(value).toEqual(serializedOk("01 12"))
    })

    test("should serialize a buffer with 128 bytes", ({ expect }) => {
      const value = schema.serialize(sampleBuffer.slice(0, 128))
      expect(value).toMatchSnapshot()
    })

    test("should reject a non-canonical length prefix which could fit one byte when allowNoncanonical is false", ({
      expect,
    }) => {
      const value = schema.parse(hexToUint8Array("81 01 12"))
      expect(value).toMatchSnapshot("error")
    })

    test("should accept a non-canonical length prefix which could fit one byte when allowNoncanonical is true", ({
      expect,
    }) => {
      const value = schema.parse(hexToUint8Array("81 01 12"), 0, {
        allowNoncanonical: true,
      })
      expect(value).toEqual(parsedOk(3, hexToUint8Array("12")))
    })

    test("should reject a non-canonical length prefix which could fit two bytes when allowNoncanonical is false", ({
      expect,
    }) => {
      const testVector = sampleBuffer.slice(0, 131)
      testVector.set(hexToUint8Array("82 00 80"), 0)
      const value = schema.parse(testVector)
      expect(value).toMatchSnapshot("error")
    })

    test("should reject a non-canonical length prefix which could fit two bytes when allowNoncanonical is true", ({
      expect,
    }) => {
      const testVector = sampleBuffer.slice(0, 131)
      testVector.set(hexToUint8Array("82 00 80"), 0)
      const value = schema.parse(testVector, 0, { allowNoncanonical: true })
      expect(value).toMatchSnapshot()
    })
  })
})
