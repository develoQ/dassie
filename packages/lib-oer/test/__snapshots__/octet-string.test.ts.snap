// Vitest Snapshot v1

exports[`octetString > with fixed length > should refuse to serialize a value of the wrong length 1`] = `
{
  "failure": [SerializeError: Expected octet string of length 22, but got 21],
  "success": false,
}
`;

exports[`octetString > with variable length > should reject a non-canonical length prefix which could fit one byte when allowNoncanonical is false > error 1`] = `
{
  "failure": [ParseError: non-canonical encoding - length prefix is not minimal],
  "success": false,
}
`;

exports[`octetString > with variable length > should reject a non-canonical length prefix which could fit two bytes when allowNoncanonical is false > error 1`] = `
{
  "failure": [ParseError: non-canonical encoding - length prefix is not minimal],
  "success": false,
}
`;

exports[`octetString > with variable length > should reject a non-canonical length prefix which could fit two bytes when allowNoncanonical is true 1`] = `
{
  "length": 131,
  "success": true,
  "value": Uint8Array [
    113,
    189,
    198,
    177,
    102,
    56,
    3,
    245,
    58,
    46,
    29,
    219,
    55,
    0,
    75,
    47,
    254,
    176,
    89,
    191,
    152,
    111,
    13,
    5,
    113,
    126,
    140,
    15,
    215,
    210,
    250,
    176,
    8,
    175,
    109,
    10,
    170,
    29,
    202,
    162,
    247,
    125,
    8,
    48,
    216,
    209,
    143,
    109,
    138,
    161,
    102,
    164,
    152,
    32,
    38,
    171,
    175,
    50,
    167,
    149,
    137,
    158,
    66,
    60,
    237,
    138,
    9,
    47,
    51,
    129,
    11,
    79,
    22,
    118,
    75,
    114,
    31,
    125,
    205,
    83,
    33,
    184,
    237,
    231,
    141,
    0,
    230,
    28,
    210,
    90,
    132,
    31,
    74,
    64,
    76,
    117,
    117,
    156,
    73,
    144,
    177,
    162,
    43,
    143,
    243,
    83,
    39,
    125,
    17,
    219,
    156,
    161,
    173,
    128,
    68,
    6,
    89,
    240,
    138,
    1,
    219,
    167,
    146,
    201,
    9,
    0,
    0,
    0,
  ],
}
`;

exports[`octetString > with variable length > should serialize a buffer with 128 bytes 1`] = `
{
  "success": true,
  "value": Uint8Array [
    129,
    128,
    221,
    141,
    113,
    113,
    189,
    198,
    177,
    102,
    56,
    3,
    245,
    58,
    46,
    29,
    219,
    55,
    0,
    75,
    47,
    254,
    176,
    89,
    191,
    152,
    111,
    13,
    5,
    113,
    126,
    140,
    15,
    215,
    210,
    250,
    176,
    8,
    175,
    109,
    10,
    170,
    29,
    202,
    162,
    247,
    125,
    8,
    48,
    216,
    209,
    143,
    109,
    138,
    161,
    102,
    164,
    152,
    32,
    38,
    171,
    175,
    50,
    167,
    149,
    137,
    158,
    66,
    60,
    237,
    138,
    9,
    47,
    51,
    129,
    11,
    79,
    22,
    118,
    75,
    114,
    31,
    125,
    205,
    83,
    33,
    184,
    237,
    231,
    141,
    0,
    230,
    28,
    210,
    90,
    132,
    31,
    74,
    64,
    76,
    117,
    117,
    156,
    73,
    144,
    177,
    162,
    43,
    143,
    243,
    83,
    39,
    125,
    17,
    219,
    156,
    161,
    173,
    128,
    68,
    6,
    89,
    240,
    138,
    1,
    219,
    167,
    146,
    201,
    9,
  ],
}
`;
